File: Data Structures & Algorithms/sliding-window-maximum/submission-1.cpp
// We'll use a heap + lazy delete to maintain an ordering
// on our window.

class Solution 
{
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) 
    {
        std::vector<int> res;
        std::priority_queue<std::pair<int, int>> max_heap;
        std::unordered_set<int> removed_set;

        for (int i = 0; i < k; i++)
        {
            max_heap.push({nums[i], i});
        }
        res.push_back(max_heap.top().first);

        int len = static_cast<int>(nums.size());
        for (int i = 0; i < len - k; i++)
        {
            int ele_to_add = nums[i + k];

            max_heap.push({ele_to_add, i + k});
            removed_set.insert(i);

            // Make sure our top element didn't get removed
            while (removed_set.contains(max_heap.top().second))
            {
                max_heap.pop();
            }

            res.push_back(max_heap.top().first);
        }

        return res;
    }
};

File: Data Structures & Algorithms/merge-strings-alternately/submission-1.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        first, second = 0, 0
        res = []

        while first < len(word1) and second < len(word2):
            res.append(word1[first])
            res.append(word2[second])
            first += 1
            second += 1
        
        if first == len(word1):
            res.append(word2[second:])
        else:
            res.append(word1[first:])

        return "".join(res)